gitlab/resource_gitlab_project.go

```diff
@@ -915,129 +915,142 @@ func resourceGitlabProjectDelete(ctx context.Context, d *schema.ResourceData, me
 func editOrAddPushRules(ctx context.Context, client *gitlab.Client, projectID string, d *schema.ResourceData) error {
 	log.Printf("[DEBUG] Editing push rules for project %q", projectID)
 
-	editOptions := expandEditProjectPushRuleOptions(d)
-	_, _, err := client.Projects.EditProjectPushRule(projectID, editOptions, gitlab.WithContext(ctx))
-	if err == nil {
-		return nil
-	}
+	pushRules, _, err := client.Projects.GetProjectPushRules(d.Id(), gitlab.WithContext(ctx))
+	// NOTE: push rules id `0` indicates that there haven't been any push rules set.
+	if err != nil || is404(err) || pushRules.ID == 0 {
+		if addOptions, needToAdd := expandAddProjectPushRuleOptions(d); needToAdd {
```

Comment:
Instead of returning `needToAdd`, you can compare the result `addOptions` with an empty struct in order to detect if it has fields set.

i.e. https://github.com/gitlabhq/terraform-provider-gitlab/blob/289199868e15d647a08f71096a38983ac1c3bacd/gitlab/resource_gitlab_project.go#L635

Reply:
Good question 😄  The thing is that the defaults in the provider (struct) not always match the defaults on the GitLab server, because admins have the possibility to set instance-wide default push rules. And because we can't post an empty body for the push rule endpoint, we have to check if the user specified a value or not (which is not supported by the SDK v2, but only the upcoming framework). This is the only "sane" way I found at the moment to achieve this.

Reply:
I think my comment stands though. `needToAdd` can be calculated because all of the fields of `gitlab.EditProjectOptions` are pointers, so you can easily detect if any of the fields have been set, by comparing it with a zero-value `gitlab.EditProjectOptions`.

Reply:
True, ... pointers 😄  I've changed it!